src/libs/antares/study/parts/hydro/finallevelinflowsmodifyer.cpp

```diff
@@ -0,0 +1,175 @@
+{
+    double reservoirCapacity = hydro_.reservoirCapacity;
+    if (-deltaReservoirLevel_ * reservoirCapacity > totalYearInflows)
+    {
+        logs.error() << "Year: " << year + 1 << ". Area: " << areaName_
+                     << ". Incompatible total inflows: " << totalYearInflows
+                     << " with initial: " << initialReservoirLevel_
+                     << " and final: " << finalReservoirLevel_ << " reservoir levels.";
+        return false;
+    }
+    return true;
+}
+
+bool FinalLevelInflowsModifier::preCheckRuleCurves(uint year) const
+{
+    double lowLevelLastDay  = hydro_.reservoirLevel[Data::PartHydro::minimum][DAYS_PER_YEAR - 1];
+    double highLevelLastDay = hydro_.reservoirLevel[Data::PartHydro::maximum][DAYS_PER_YEAR - 1];
+
+    if (finalReservoirLevel_ < lowLevelLastDay || finalReservoirLevel_ > highLevelLastDay)
+    {
+        logs.error() << "Year: " << year + 1 << ". Area: " << areaName_
+                     << ". Specifed final reservoir level: " << finalReservoirLevel_
+                     << " is incompatible with reservoir level rule curve [" << lowLevelLastDay
+                     << " , " << highLevelLastDay << "]";
+        return false;
+    }
+    return true;
+}
+
+void FinalLevelInflowsModifier::initialize(const Matrix<double>& scenarioInitialHydroLevels,
+                                           const Matrix<double>& scenarioFinalHydroLevels,
+                                           const uint lastSimulationDay,
+                                           const uint nbYears)
+{
+    isApplicable_.assign(nbYears, false);
+    deltaLevel.assign(nbYears, 0.);
+    InitialLevels_ = &(scenarioInitialHydroLevels.entry[areaIndex_]);
+    FinalLevels_ = &(scenarioFinalHydroLevels.entry[areaIndex_]);
+    lastSimulationDay_ = lastSimulationDay;
+}
+
+bool FinalLevelInflowsModifier::isActive()
+{
+    return  hydro_.reservoirManagement && 
+            !hydro_.useWaterValue &&
+            isValidLevel(finalReservoirLevel_) &&
+            isValidLevel(initialReservoirLevel_);
+}
+
+bool FinalLevelInflowsModifier::makeChecks(uint year)
+{
+    // Simulation must end on day 365 and reservoir level must be initiated in January
+    bool checksOk = preCheckStartAndEndSim(year);
+
+    // Reservoir_levelDay_365 – reservoir_levelDay_1 ≤ yearly_inflows
+    double totalInflows = calculateTotalInflows(year);
+    checksOk = preCheckYearlyInflow(totalInflows, year) && checksOk;
+
+    // Final reservoir level set by the user is within the
+    // rule curves for the final day
+    checksOk = preCheckRuleCurves(year) && checksOk;
+
+    isApplicable_.at(year) = checksOk;
+
+    return checksOk;
+}
+
+bool FinalLevelInflowsModifier::isApplicable(uint year)
+{
+    // If isApplicable_.size() == 0, then instance was not properly initialized
+    // and is not applicable.
+    return isActive() && isApplicable_.size() && isApplicable_.at(year);
```

Comment:
Calling method isActive() works quite ok in CheckInfeasibility.
However, calling isActive() in method isApplicable() (that is only called in HydroManagement method changeInflowsToAccommodateFinalLevels) is redundant and even more dangerous.
Reason -  isActive() is called with stale data for finalReservoirLevel_ & initialReservoirLevel_ ( last one - from the last year). 
So, if we have 10 MCy. Let's assume that the logic should be implemented for the first 9 years. And for the last (10th) year logic should not be implemented because of the invalid init/final reservoir levels. What will actually happen. The logic implementation will be skipped for all 10 years.

Reply:
I totally agree.